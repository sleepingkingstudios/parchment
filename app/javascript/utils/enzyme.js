import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
