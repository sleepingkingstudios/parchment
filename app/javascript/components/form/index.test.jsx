import React from 'react';
import { shallow } from 'enzyme';

import Form from './index';

const MockInput = () => (<span />);

describe('<Form />', () => {
  const children = (<MockInput />);
  const onSubmitAction = jest.fn(() => ({ payload: { ok: true } }));
  const form = {
    data: {},
    onChangeAction: jest.fn(),
    onSubmitAction,
  };
  const defaultProps = { children, form };

  it('should wrap the contents in a form tag', () => {
    const rendered = shallow(<Form {...defaultProps} />);

    expect(rendered).toHaveDisplayName('form');
  });

  it('should render the children', () => {
    const rendered = shallow(<Form {...defaultProps} />);
    const input = rendered.find('MockInput');

    expect(input).toExist();
  });

  it('should render the hidden submit input', () => {
    const rendered = shallow(<Form {...defaultProps} />);
    const input = rendered.find('input');
    const type = 'submit';

    expect(input).toExist();
    expect(input).toHaveClassName('d-none');
    expect(input).toHaveProp({ type });
  });

  it('should set the onSubmit event handler', () => {
    const rendered = shallow(<Form {...defaultProps} />);
    const { onSubmit } = rendered.props();
    const preventDefault = jest.fn();
    const event = { preventDefault };

    expect(typeof onSubmit).toEqual('function');

    const action = onSubmit(event);

    expect(preventDefault).toHaveBeenCalled();
    expect(onSubmitAction).toHaveBeenCalledWith();
    expect(action).toEqual({ payload: { ok: true } });
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<Form {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });

  describe('with className: value', () => {
    const props = { ...defaultProps, className: 'example-form' };

    it('should wrap the contents in a form tag', () => {
      const rendered = shallow(<Form {...props} />);
      const { className } = props;

      expect(rendered).toHaveDisplayName('form');
      expect(rendered).toHaveClassName(className);
    });
  });
});
