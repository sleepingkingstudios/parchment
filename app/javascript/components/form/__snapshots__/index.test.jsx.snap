// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should match the snapshot 1`] = `
<form
  className={null}
  onSubmit={[Function]}
>
  <input
    className="d-none"
    onClick={[Function]}
    type="submit"
  />
  <MockInput />
</form>
`;
